File: Fotty/Design/TeamColors.swift
import SwiftUI

// Hex color parser for SwiftUI
extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (1, 1, 1, 1)
        }

        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue: Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
}

enum TeamColorResolver {
    static let teamColors: [String: Color] = [
        // National teams
        "argentina": Color(hex: "#74acdf"),
        "egypt": Color(hex: "#ce1126"),
        "switzerland": Color(hex: "#da291c"),
        "colombia": Color(hex: "#fcd116"),
        "france": Color(hex: "#002395"),
        "morocco": Color(hex: "#c1272d"),
        "mexico": Color(hex: "#006847"),
        "south africa": Color(hex: "#007a4d"),
        "south korea": Color(hex: "#cd1125"),
        "korea": Color(hex: "#cd1125"),
        "czech": Color(hex: "#11457e"),
        "canada": Color(hex: "#ff0000"),
        "bosnia": Color(hex: "#002f6c"),
        "usa": Color(hex: "#002868"),
        "united states": Color(hex: "#002868"),
        "netherlands": Color(hex: "#ff4f00"),
        "belgium": Color(hex: "#e30613"),
        "brazil": Color(hex: "#ffdf00"),
        "germany": Color(hex: "#ffffff"),
        "england": Color(hex: "#ce1124"),
        "italy": Color(hex: "#113a5d"),
        "spain": Color(hex: "#c60b1e"),
        "portugal": Color(hex: "#ff0000"),
        "croatia": Color(hex: "#ff0000"),
        "uruguay": Color(hex: "#0081c6"),
        "japan": Color(hex: "#bc002d"),
        "australia": Color(hex: "#00008b"),

        // Clubs
        "arsenal": Color(hex: "#ef0107"),
        "chelsea": Color(hex: "#034694"),
        "barcelona": Color(hex: "#004d98"),
        "real madrid": Color(hex: "#8a9bb8"),
        "liverpool": Color(hex: "#c8102e"),
        "manchester united": Color(hex: "#da291c"),
        "manchester city": Color(hex: "#6cabdd"),
        "bayern munich": Color(hex: "#dc052d"),
        "juventus": Color(hex: "#7d7d7d"),
        "inter milan": Color(hex: "#0068a8"),
        "tottenham": Color(hex: "#132257"),
        "paris saint-germain": Color(hex: "#002F6C"),
        "psg": Color(hex: "#002F6C"),
        "dortmund": Color(hex: "#FDE100"),
        "milan": Color(hex: "#E30613"),
        "ac milan": Color(hex: "#E30613"),
        "atletico madrid": Color(hex: "#CB3524"),
        "ajax": Color(hex: "#D2122E")
    ]

    private static var resolvedCache: [String: Color] = [:]
    private static let lock = NSLock()

    static func resolve(teamName: String) -> Color? {
        let name = teamName.lowercased().trimmingCharacters(in: .whitespacesAndNewlines)
        guard !name.isEmpty else { return nil }

        lock.lock()
        if let cached = resolvedCache[name] {
            lock.unlock()
            return cached
        }
        lock.unlock()

        for (key, color) in teamColors {
            if name.contains(key) {
                lock.lock()
                resolvedCache[name] = color
                lock.unlock()
                return color
            }
        }
        
        // Deterministic fallback: Hash the name to generate a consistent, vibrant color (HSL style)
        var hash: UInt64 = 5381
        for char in name.utf8 {
            hash = ((hash << 5) &+ hash) &+ UInt64(char)
        }
        
        // Map hash to a hue value between 0.0 and 1.0
        let hue = Double(hash % 360) / 360.0
        let color = Color(hue: hue, saturation: 0.85, brightness: 0.95)
        lock.lock()
        resolvedCache[name] = color
        lock.unlock()
        return color
    }
}

File: web/src/lib/v2/team-colors.ts
// Shared preset colors for major football clubs and national sides.
export const TEAM_COLORS: Record<string, { primary: string; secondary: string }> = {
  // National sides
  "argentina": { primary: "#74acdf", secondary: "#ffffff" },
  "egypt": { primary: "#ce1126", secondary: "#000000" },
  "switzerland": { primary: "#da291c", secondary: "#ffffff" },
  "colombia": { primary: "#fcd116", secondary: "#003893" },
  "france": { primary: "#002395", secondary: "#ffffff" },
  "morocco": { primary: "#c1272d", secondary: "#006233" },
  "mexico": { primary: "#006847", secondary: "#ce1126" },
  "south africa": { primary: "#007a4d", secondary: "#ffb612" },
  "south korea": { primary: "#cd1125", secondary: "#0a1b3a" },
  "korea": { primary: "#cd1125", secondary: "#0a1b3a" },
  "czech": { primary: "#11457e", secondary: "#d90f17" },
  "canada": { primary: "#ff0000", secondary: "#ffffff" },
  "bosnia": { primary: "#002f6c", secondary: "#febd11" },
  "usa": { primary: "#002868", secondary: "#bf0a30" },
  "united states": { primary: "#002868", secondary: "#bf0a30" },
  "netherlands": { primary: "#ff4f00", secondary: "#002147" },
  "belgium": { primary: "#e30613", secondary: "#ffd919" },
  "brazil": { primary: "#ffdf00", secondary: "#009b3a" },
  "germany": { primary: "#ffffff", secondary: "#d90000" },
  "england": { primary: "#ce1124", secondary: "#ffffff" },
  "italy": { primary: "#113a5d", secondary: "#008c45" },
  "spain": { primary: "#c60b1e", secondary: "#ffc400" },
  "portugal": { primary: "#ff0000", secondary: "#006600" },
  "croatia": { primary: "#ff0000", secondary: "#002f6c" },
  "uruguay": { primary: "#0081c6", secondary: "#ffffff" },
  "japan": { primary: "#bc002d", secondary: "#002f6c" },
  "australia": { primary: "#00008b", secondary: "#ffd700" },

  // Clubs
  "arsenal": { primary: "#ef0107", secondary: "#063672" },
  "chelsea": { primary: "#034694", secondary: "#ee242c" },
  "barcelona": { primary: "#004d98", secondary: "#a50044" },
  "real madrid": { primary: "#8a9bb8", secondary: "#121f4c" },
  "liverpool": { primary: "#c8102e", secondary: "#f6eb61" },
  "manchester united": { primary: "#da291c", secondary: "#000000" },
  "manchester city": { primary: "#6cabdd", secondary: "#1c2c5b" },
  "bayern munich": { primary: "#dc052d", secondary: "#0066b2" },
  "juventus": { primary: "#7d7d7d", secondary: "#000000" },
  "inter milan": { primary: "#0068a8", secondary: "#221f1f" },
  "tottenham": { primary: "#132257", secondary: "#ffffff" },
  "paris saint-germain": { primary: "#002F6C", secondary: "#E30613" },
  "psg": { primary: "#002F6C", secondary: "#E30613" },
  "dortmund": { primary: "#FDE100", secondary: "#000000" },
  "milan": { primary: "#E30613", secondary: "#000000" },
  "ac milan": { primary: "#E30613", secondary: "#000000" },
  "atletico madrid": { primary: "#CB3524", secondary: "#192C5B" },
  "ajax": { primary: "#D2122E", secondary: "#ffffff" },
};

export function resolveTeamColor(teamName?: string): string | null {
  if (!teamName) return null;
  const name = teamName.toLowerCase().trim();
  for (const [key, colors] of Object.entries(TEAM_COLORS)) {
    if (name.includes(key)) return colors.primary;
  }
  return null;
}
